test/ViewExtensionLibraryTests/LibraryResourceProviderTests.cs

```diff
@@ -305,6 +305,68 @@ public void PackagedCustomNodeSearchElementLoadedType()
             Assert.AreEqual(url.Url, item.iconUrl);
         }
 
+        [Test]
+        [Category("UnitTests")]
+        public void CustomNodePropertiesWindowValidateCategories()
+        {
+            // sample CN #1
+            var CN1_Info = new CustomNodeInfo(
+                Guid.NewGuid(), 
+                "category test node 1", 
+                "base.level1.level2.level3",
+                "A node for testing CN dialog categories drop - down",
+                @"C:\temp\category_test_node_1.dyf");
+
+            // sample CN #2
+            var CN2_Info = new CustomNodeInfo(
+                Guid.NewGuid(),
+                "category test node 2",
+                "base.level1.level2.level3.level4.level5",
+                "A node for testing CN dialog categories drop - down",
+                @"C:\temp\category_test_node_2.dyf");
+
+            var CN1_Mock = new Mock<ICustomNodeSource>();
+            var CN2_Mock = new Mock<ICustomNodeSource>();
+
+            var CN1_element = new CustomNodeSearchElement(CN1_Mock.Object, CN1_Info);
+            var CN2_element = new CustomNodeSearchElement(CN2_Mock.Object, CN2_Info);
+
+            var provider = new NodeItemDataProvider(new NodeSearchModel());
+            var CN1_LoadedType = provider.CreateLoadedTypeItem<LoadedTypeItem>(CN1_element);
+            var CN2_LoadedType = provider.CreateLoadedTypeItem<LoadedTypeItem>(CN2_element);
+
+            var elements = new NodeSearchElement[] { CN1_element, CN2_element };
+
+            // Call function used to populate Add-ons categories drop-down in CN properties window
+            List<string> addOnCategories = Dynamo.Controls.DynamoView.getUniqueAddOnCategories(elements).ToList();
+
+            // Expected results
+            var CN1_expectedQualifiedName = "dyf://base.level1.level2.level3.category test node 1";
+            var CN2_expectedQualifiedName = "dyf://base.level1.level2.level3.level4.level5.category test node 2";
+
+            // Expected unique categories that will be populated for the drop-down in the function above
+            string[] expectedCategories = new string[]
+            {
+                "base",
+                "base.level1",
+                "base.level1.level2",
+                "base.level1.level2.level3",
+                "base.level1.level2.level3.level4",
+                "base.level1.level2.level3.level4.level5"
+                // Should not include node names at bottom levels
+            };
+
+            // Verify expected results
+            Assert.AreEqual(CN1_expectedQualifiedName, CN1_LoadedType.fullyQualifiedName);
+            Assert.AreEqual(CN2_expectedQualifiedName, CN2_LoadedType.fullyQualifiedName);
+            Assert.AreEqual(addOnCategories.Count, expectedCategories.Length);
+
+            for (int i = 0; i < addOnCategories.Count; i++)
+            {
+                Assert.AreEqual(addOnCategories[i], expectedCategories[i]);
+            }
+        }
+
         [Test]
```

Comment:
@mjkkirschner updated this test with an additional CN to verify we aren't getting any duplicate results

Reply:
@alfarok how do we make sure the two identical dyf are sure to be ported to C:\temp\category_test_node_2.dyf ahead of the test? Are they used in other tests?

Reply:
I dont think theres anything there at all, these are just mocks to test this function. - right @alfarok

Reply:
@QilongTang The .dyfs are actually slightly different but they are only used in this test.  The test actually doesn't require them to be present in the library it just verifies the function (`getUniqueAddOnCategories`) used to populate the categories dropdown is returning the expected results.